### PR TITLE
Better --port & --bridge-port options validation in CLI

### PR DESCRIPTION
**Pull request by KtorZ**
# Issue Number

#368 

# Overview

- [ ] I have improved validations for `--port` and `--bridge` options to require a value within reasonable bounds.

# Comments

### CODE REVIEW

File: lib/cli/src/Cardano/CLI.hs [hunk 1]
```diff
@@ -104,11 +113,32 @@ import qualified Data.Text.IO as TIO
...
+    deriving stock (Eq, Show, Generic)
+    deriving newtype (Enum, Ord)
+
+-- NOTE
+-- TCP port ranges from [[-1;65535]] \ {0}
+-- However, ports in [[-1; 1023]] \ {0} are well-known ports reserved
+-- and only "bindable" through root privileges.
+--
+-- Ports above 49151 are also reserved for ephemeral connections. This
+-- leaves us with a valid range of [[1024;49151]] for TCP ports.
```
Comment (piotr-iohk) [inline lib/cli/src/Cardano/CLI.hs:126]: :+1:

File: lib/core/test/integration/Test/Integration/Scenario/CLI/Port.hs [hunk 2]
```diff
@@ -51,6 +54,30 @@ import Test.Integration.Framework.DSL
...
+        let tests =
+                [ ("serve", "--port", getPort minBound - 14)
+                , ("serve", "--port", getPort maxBound + 14)
+                , ("serve", "--bridge-port", getPort minBound - 14)
+                , ("serve", "--bridge-port", getPort maxBound + 14)
+                , ("launch", "--port", getPort minBound - 14)
+                , ("launch", "--port", getPort maxBound + 14)
+                , ("launch", "--bridge-port", getPort minBound - 14)
+                , ("launch", "--bridge-port", getPort maxBound + 14)
+                ]
```
Comment (piotr-iohk) [inline lib/core/test/integration/Test/Integration/Scenario/CLI/Port.hs:69]: That's nice (btw. any reason for `14`?). I'd also add more negative and positive cases for some values that usually tend to be _suspicious_:
 - boundary values (here `minBound` and `maxBound` and their nearest vicinity (`_Bound + 1`, `_Bound - 1`))
 - arbitrary string
 - negative value
 - `0`
 - ...

Comment (KtorZ) [reply lib/core/test/integration/Test/Integration/Scenario/CLI/Port.hs:69]: No reason actually, it's completely arbitrary. I'd suggest that we could add more tests like this at the `unit` level, and not in integration tests, as they would be less costly to run. The above make sure that the parser gets called, and we can extensively test the parser in unit tests, what do you think ?

Comment (KtorZ) [reply lib/core/test/integration/Test/Integration/Scenario/CLI/Port.hs:69]: See: https://github.com/input-output-hk/cardano-wallet/pull/425/files#diff-7e848cbd8fd36a221b008a73bd1e550aR55

Comment (KtorZ) [reply lib/core/test/integration/Test/Integration/Scenario/CLI/Port.hs:69]: Also, I've made the above more arbitrary to make it clear that there's nothing special about them.

Comment (piotr-iohk) [reply lib/core/test/integration/Test/Integration/Scenario/CLI/Port.hs:69]: That's ok, just want to point out that we had some cases previously where parser that worked fine on `unit` level didn't work as expected at `integration` level (address-pool-gap = -1000 was 24). Therefore, in principle, I'd prefer having few more _classic_ black box cases like that on integration level as well. Ofc the risk here maybe unlikely, and potential gain from lest costly test is better (on the other hand who would've known that -1000 is 24 ¯\_(ツ)_/¯ ;) )

**Review by piotr-iohk — APPROVED** [review]
lgtm :+1:

File: lib/core/test/integration/Test/Integration/Scenario/CLI/Port.hs (hunk 2 above)
Comment (KtorZ) [reply lib/core/test/integration/Test/Integration/Scenario/CLI/Port.hs:69]: Well... the parser wasn't actually fine on the unit level either in that case :grimacing: ... because the same code is used in both the API and the primitive types. Generally speaking however, I do agree about the approach on integration tests. Yet here, I think we've a nice trade-off -> we know the same parser is being used (cf error message, I know it's a poor guarantee but well), and we have it well-tested at the unit-level.

Comment (jonathanknowles) [reply lib/core/test/integration/Test/Integration/Scenario/CLI/Port.hs:69]: If these numbers are all arbitrarily chosen, then I would suggest adding a comment to that effect. For example:
```hs
        let tests =
                -- The numbers below are all chosen arbitrarily:
                [ ("serve", "--port", getPort minBound - 14)
                , ("serve", "--port", getPort maxBound + 42)
                , ("serve", "--bridge-port", getPort minBound - 6)
                , ("serve", "--bridge-port", getPort maxBound + 523)
                , ("launch", "--port", getPort minBound - 1)
                , ("launch", "--port", getPort maxBound + 59375)
                , ("launch", "--bridge-port", getPort minBound - 5621)
                , ("launch", "--bridge-port", getPort maxBound + 1)
                ]
```

**Comment by KtorZ** [comment]
I've merged a few PRs that were already reviewed & accepted "by-hand" into this PR to batch a bit PR / CI builds since we're giving CI a very hard time today after trying to solve whatever happened with travis.

Also added a commit to fix buildkite and re-generate whatever was needed for nix. Git history now looks like what follows:

```
* 32f96093 - (HEAD -> KtorZ/--port-validation, origin/KtorZ/--port-validation) re-generate nix machinery (52 seconds ago) <KtorZ>
*   38d11c12 - Merge branch 'jonathanknowles/severity-pass-through' into KtorZ/--port-validation (52 seconds ago) <KtorZ>
|\  
| * 89977898 - Pass CLI argument `--min-log-severity` through when calling `serve` from `launch`. (52 seconds ago) <Jonathan Knowles>
|/  
*   08fe8c65 - Merge branch 'jonathanknowles/wallet-server-startup-log-entry' into KtorZ/--port-validation (10 minutes ago) <KtorZ>
|\  
| * ca2f088b - Write version string to the log when starting the wallet server. (10 minutes ago) <Jonathan Knowles>
|/  
* df3cfbb1 - add more negative unit test cases for port decoding (16 minutes ago) <KtorZ>
* 232ae48b - Make port numbers a bit more arbitrary in tests (16 minutes ago) <KtorZ>
* 1bcef874 - re-generate nix machinery (16 minutes ago) <KtorZ>
* 37250bb8 - constraints 'Port' type between valid bounds (16 minutes ago) <KtorZ>
*   43e49aa8 - (origin/master) Merge pull request #413 from input-output-hk/piotr/testing (17 minutes ago) <Matthias Benkort>
```
